nsmdp/nsmdp.go

```diff
@@ -112,15 +146,111 @@ func (n *NSMDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.Allocate
 	return &responses, nil
 }
 
-func (n *NSMDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+func (n *nsmClientEndpoints) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	n.logger.Infof("ListAndWatch was called with s: %+v", s)
 	for {
-		select {
-		case <-n.stop:
-			return nil
-		case dev := <-n.updatedevs:
-			n.devs[dev.ID] = dev
-			devs := []*pluginapi.Device{dev.Device}
-			s.Send(&pluginapi.ListAndWatchResponse{Devices: devs})
+		resp := new(pluginapi.ListAndWatchResponse)
+		for _, dev := range n.nsmSockets {
+			resp.Devices = append(resp.Devices, dev.device)
 		}
+		if err := s.Send(resp); err != nil {
+			n.logger.Errorf("Failed to send response to kubelet: %v\n", err)
+		}
+		time.Sleep(5 * time.Second)
 	}
+	return nil
+}
+
+func (n *nsmClientEndpoints) PreStartContainer(context.Context, *pluginapi.PreStartContainerRequest) (*pluginapi.PreStartContainerResponse, error) {
+	n.logger.Infof("PreStartContainer was called.")
+	return &pluginapi.PreStartContainerResponse{}, nil
+}
+
+// Register registers
+func Register(kubeletEndpoint string) error {
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	defer conn.Close()
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot connect to kubelet service: %v", err)
+	}
+	client := pluginapi.NewRegistrationClient(conn)
+	reqt := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     serverSock,
+		ResourceName: resourceName,
+	}
+
+	_, err = client.Register(context.Background(), reqt)
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot register to kubelet service: %v", err)
+	}
+	return nil
+}
+
+func dial(ctx context.Context, unixSocketPath string) (*grpc.ClientConn, error) {
+	c, err := grpc.DialContext(ctx, unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func startDeviceServer(nsm *nsmClientEndpoints) error {
+	// Initial socket clean up
+	listenEndpoint := path.Join(pluginapi.DevicePluginPath, serverSock)
+	fi, err := os.Stat(listenEndpoint)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(listenEndpoint)
```

Comment:
Do we want to drop the error in os.Remove?

Reply:
I can add error handling here.